README.md
# Build Webex ChatBots in JavaScript

Yet [another opiniated framework](https://github.com/CiscoDevNet/awesome-webex#bot-frameworks) to build [Webex Teams Bots](https://developer.webex.com/bots.html) in Node.js:
- simple design to both learn and experiment Webhooks concepts in a snatch,
- flexibility to let your bot listen to raw Webhook events, or directly respond to commands,  
- supports [Webex Cards](https://github.com/CiscoDevNet/node-sparkbot#capture-inputs-submitted-via-cards) and [webhook check (creation/update)](https://github.com/CiscoDevNet/node-sparkbot#auto-register-webhooks),
- leveraged by a few [DevNet learning labs](https://learninglabs.cisco.com/tracks/collab-cloud/spark-apps/collab-spark-botl-ngrok/step/1), and [bot samples](https://github.com/CiscoDevNet/node-sparkbot-samples). 

This project focusses on the [framework itself](#architecture) and its [testing companions](./tests/README.md).

If you're looking for ready-to-run Chatbots built with the library, jump to the [node-sparkbot-samples repo](https://github.com/CiscoDevNet/node-sparkbot-samples).


## Quickstart

Copy a sample from [Quickstart](quickstart/).

For **Mac, Linux and bash users**, open a terminal and type:

```shell
git clone https://github.com/CiscoDevNet/node-sparkbot
cd node-sparkbot
npm install
DEBUG=sparkbot* node tests/onEvent-all-all.js
```

For **Windows users**, open a command shell and type:

```shell
git clone https://github.com/CiscoDevNet/node-sparkbot
cd node-sparkbot
npm install
set DEBUG=sparkbot*
node tests/onEvent-all-all.js
```

**Done, your bot is live**

Let's check it's live by hitting its healthcheck endpoint:

```
# simply run: curl http://localhost:8080 
# or if you like formatting, install jq and run:
$ curl http://localhost:8080 | jq -C
{
  "message": "Congrats, your Webex Teams webhook is up and running",
  "since": "2016-09-23T07:46:52.397Z",
  "tip": "Register your bot as a WebHook to start receiving events: https://developer.webex.com/endpoint-webhooks-post.html",
  "listeners": [
    "messages/created"
  ],
  "token": false,
  "account": {},
  "interpreter": {},
  "commands": [],
}
```

**Congrats, your bot is now up and running**

Now, let's make Webex post events to our bot.
- if your bot is running on a local machine, you need to [expose your bot to the internet](docs/SettingUpYourSparkBot.md#expose-you-bot).
- and lastly, [register your bot by creating a Webhook](docs/SettingUpYourSparkBot.md#register-your-bot-as-a-spark-webhook).

Note that if you want your bot to respond to commands, add a Webex Teams API token on the command line (see below).

Finally, we suggest you take a look at the [tests](tests/README.md) as they provide a great way to discover the framework features.


### Respond to commands

At startup, the library looks for the ACCESS_TOKEN environment variable.
> Note that ACCESS_TOKEN is still accepted but deprecated as of v1.x of the framework

If present, the library will leverage the token to retreive new message contents, and automatically detect the Webex Teams account associated to the token and take initialization options to fit common scenarios, see [Account detection](#account-type-detection).

As messages flow in, the library automatically removes bot mentions when relevant, so that you can focus on the command itself.

```shell
DEBUG=sparkbot*  ACCCESS_TOKEN=Very_Secret  node tests/onCommand.js

...
  sparkbot webhook instantiated with default configuration +0ms
  sparkbot addMessagesCreatedListener: listener registered +89ms
  sparkbot bot started on port: 8080 +8ms
  sparkbot:interpreter bot account detected, name: CiscoDevNet (bot) +1s
```


### Capture inputs submitted via Cards

We'll use a sample card that proposes a single `name` entry field.

![](docs/img/card-submission.png)

To create the card, create a Webex Teams space and place the following requests with the roomId of your space, typically via Postman.
Note: you can use this [postman collection](https://www.getpostman.com/collections/1f5e101d8290a5303c90) to experiment with Cards. After importing the collection, make sure to add a BOT_TOKEN variable to your Postman environment.

```
POST https://api.ciscospark.com/v1/messages
Authorization: Bearer YOUR_BOT_TOKEN
{{
    "roomId": "{{_room}}",
    "markdown": "[Learn more](https://adaptivecards.io) about Adaptive Cards.",
    "attachments": [
        {
            "contentType": "application/vnd.microsoft.card.adaptive",
            "content": {
                "type": "AdaptiveCard",
                "version": "1.0",
                "body": [
                    {
                        "type": "TextBlock",
                        "text": "Please enter your name:"
                    },
                    {
                        "type": "Input.Text",
                        "id": "name",
                        "title": "New Input.Toggle",
                        "placeholder": "name"
                    }
                ],
                "actions": [
                    {
                        "type": "Action.Submit",
                        "title": "Submit"
                    }
                ]
            }
        }
    ]
}
``` 

As cards are submitted, the [onCardSubmission()]() function invokes your custom code logic.
Note: if a PUBLIC_URL environment variable is present, the example will automatically create a Webhook.

```shell
 DEBUG=sparkbot ACCESS_TOKEN=Y2QzMz_typically_a_bot_token_JlODAzZmItYTVm PUBLIC_URL="https://d3fc85fe.ngrok.io" node tests/onCardSubmission-webhook.js

...
 sparkbot webhook instantiated with default configuration +0ms
  sparkbot bot started on port: 8080 +37ms
  sparkbot webhook already exists with same properties, no creation needed +664ms
webhook successfully checked, with id: Y2lzY29zcGFyazovL3VzL1dFQkhPT0svYTkyYWU5NjMtOTNmYS00YTE0LWEwOGItYTMzMjQ5MzA3MGQ4
  sparkbot webhook invoked +9s
  sparkbot calling listener for resource/event: attachmentActions/created, with data context: Y2lzY29zcGFyazovL3VzL0FUVEFDSE1FTlRfQUNUSU9OLzkyNmEzNDYwLWMzMjktMTFlOS05OGQyLTg5ZDBlNGQyZDU1Mg +6ms
new attachmentActions from personId: Y2lzY29zcGFyazovL3VzL1BFT1BMRS85MmIzZGQ5YS02NzVkLTRhNDEtOGM0MS0yYWJkZjg5ZjQ0ZjQ , with inputs
   name: CiscoDevNet
```


Here is the code used by the sample, hou can check the full code sample here:

```javascript
// Starts your Webhook with a default configuration where the Webex API access token is read from ACCESS_TOKEN
const SparkBot = require("../sparkbot/webhook");
const bot = new SparkBot();

// Create webhook
const publicURL = process.env.PUBLIC_URL || "https://d3fc85fe.ngrok.io";
bot.secret = process.env.WEBHOOK_SECRET || "not THAT secret";
bot.createOrUpdateWebhook("register-bot", publicURL, "attachmentActions", "created", null, bot.secret);
 
// Process new card submissions
bot.onCardSubmission(function (trigger, attachmentActions) {

   console.log(`new attachmentActions from personId: ${trigger.data.personId} , with inputs`);
   Object.keys(attachmentActions.inputs).forEach(prop => {
      console.log(`   ${prop}: ${attachmentActions.inputs[prop]}`);
   });
});
```


## Architecture

The library supports the full set of Webex Teams webhooks, see https://developer.webex.com/webhooks-explained.html

As Webex fires Webhook events, the related listener functions are called.

You can register to listen to a WebHook event by calling the function **".on(<resources>,<event>)"**

Note that the library implements a shortcut that lets you listen to all Webhook events. Check sample code [onEvent-all-all.js](tests/onEvent-all-all.js).

```javascript
var SparkBot = require("sparkbot");
var bot = new SparkBot();
 
bot.onEvent("all", "all", function(trigger) {
  
    // YOUR CODE HERE
    console.log("EVENT: " + trigger.resource + "/" + trigger.event + ", with data id: " + trigger.data.id + ", triggered by person id:" + trigger.actorId);
});
```

This other example code [onEvent-messages-created.js](tests/onEvent-messages-created.js) illustrates how to listen to (Messages/Created) Webhook events.

```javascript
// Starts your Webhook with default configuration where the Webex Teams API access token is read from the ACCESS_TOKEN env variable 
var SparkBot = require("sparkbot");
var bot = new SparkBot();

bot.onEvent("messages", "created", function(trigger) {
    console.log("new message from: " + trigger.data.personEmail + ", in room: " + trigger.data.roomId);
    bot.decryptMessage(trigger, function (err, message) {
        if (err) {
            console.log("could not fetch message contents, err: " + err.message); 
            return;
        }

        // YOUR CODE HERE
        console.log("processing message contents: " + message.text);
    });
});
```


### New message listener

The library also provides a shortcut easy way to listen to only new messages, via the .onMessage() function.

Note that this function is not only a shorcut to create a **".on('messages', 'created')"** listener.
It also automatically fetches for you the text of the new message by requesting Webex Teams for the message details.
As the message is fetched behind the scene, you should position a ACCESS_TOKEN env variable when starting up your bot.

Check the [onMessage.js](tests/onMessage.js) for an example:

```javascript
// Starts your Webhook with default configuration where the Webex Teams API access token is read from the ACCESS_TOKEN env variable 
SparkBot = require("node-sparkbot");
var bot = new SparkBot();
 
bot.onMessage(function(trigger, message) {
 
  // ADD YOUR CUSTOM CODE HERE
  console.log("new message from: " + trigger.data.personEmail + ", text: " + message.text);
});
```

Note that most of the time, you'll want to check for the presence of a keyword to take action.
To that purpose, you can check this example: [onMessage-asCommand](tests/onMessage-asCommand.js).

```javascript
// Starts your Webhook with default configuration where the Webex Teams API access token is read from the ACCESS_TOKEN env variable 
var SparkBot = require("node-sparkbot");
var bot = new SparkBot();

bot.onMessage(function (trigger, message) {
    console.log("new message from: " + trigger.data.personEmail + ", text: " + message.text);

    var command = bot.asCommand(message);
    if (command) {
        // // ADD YOUR CUSTOM CODE HERE
        console.log("detected command: " + command.keyword + ", with args: " + JSON.stringify(command.args));
    }
});
```

_Note that The onCommand function below is pretty powerful, as it not only checks for command keywords, but also removes any mention of your bot, as this mention is a Webex pre-requisite for your bot to receive a message in a group space,
but it meaningless for your bot to process the message._

Well that said, we're ready to go thru the creation of interactive assistants.


### Interactive assistants

To implement an interative assistant, you would typically:
- respond to commands (keywords) via an `onCommand()` listener function
   - with an option to trim mention if your bot is mentionned in a group room
   - and an option to specify a fallback command
   - please check [onCommand](tests/onCommand.js) sample
- respond to attachementActions submissions (cards) via an `onCardSubmission()` listener function
   - please check [onCardSubmission](tests/onCardSubmission.js) sample
- manually create a webhook via a POST /webhooks request against the Webex REST API
   - OR use the `createOrUpdateWebhook()` function 
   - please check [onCommand-webhook](tests/onCommand-webhook.js) or [onCardSubmission-webhook](tests/onCardSubmission-webhook.js) samples


### Healthcheck

The library automatically exposes an healthcheck endpoint.
As such, hitting GET / will respond a 200 OK with an attached JSON payload.

The healcheck JSON payload will give you extra details: 
- token:true  // if a token was detected at launch
- account     // detailled info about the Webex Teams account tied to the token
- listeners   // events for which your bot has registered a listener
- commands    // commands for which your bot is ready to be activated
- interpreter // preferences for the command interpreter

   
```json
// Example of a JSON healthcheck
{
  "message": "Congrats, your Webex Teams bot is up and running",
  "since": "2016-09-01T13:15:39.425Z",
  "tip": "Register webhooks for your bot to start receiving events: https://developer.webex.com/endpoint-webhooks-post.html"
  "listeners": [
    "messages/created"
  ],
  "token": true,
  "account": {
    "type": "human",
    "person": {
      "id": "Y2lzY29zcGFyazovL3VzL1BFT1BMRS85MmIzZGQ5YS02NzVkLTRhNDEtOGM0MS0yYWJkZjg5ZjQ0ZjQ",
      "emails": [
        "stsfartz@cisco.com"
      ],
      "displayName": "Stève Sfartz",
      "avatar": "https://1efa7a94ed216783e352-c62266528714497a17239ececf39e9e2.ssl.cf1.rackcdn.com/V1~c2582d2fb9d11e359e02b12c17800f09~aqSu09sCTVOOx45HJCbWHg==~1600",
      "created": "2016-02-04T15:46:20.321Z"
    }
  },
  "interpreter": {
    "prefix": "/",
    "trimMention": true,
    "ignoreSelf": false
  },
  "commands": [
    "help"
  ]
}
``` 


### Account Type Detection

If a Webex teams API access token has been specified at launch, the library will request details about the Webex Teams account.
From the person details provided, the library will infer if the token matches a individual (HUMAN) or a bot account (MACHINE).
Because of restrictions concerning bots, the library will setup a default behavior from the account type, unless configuration parameters have already been provided at startup.

_Note that the automatic account detection procedure is done asynchronously at launch._

Here is the set of extra information and behaviors that relate to the automatic bot detection:
    - your bot is populated with an account property
    - your bot is added a nickName property
    - a type is attached to your bot instance [HUMAN | MACHINE | OTHER]


### Authenticating Requests via payload signature 

To ensure paylods received by your bots come from Webex, you can supply a secret parameter at Webhook creation.
Every payload posted to your bot will then contain an extra HTTP header "X-Spark-Signature" containing an HMAC-SHA1 signature of the payload.

Once you've created the webhook for your bot with a secret parameter, 
you can either specify a SECRET environment variable on the command line or in your code.

Command line example:
```shell
DEBUG=sparkbot*  ACCESS_TOKEN=your_bot_token WEBHOOK_SECRET=your_secret   node tests/onEvent-check-secret.js
...
```

Code example:
```javascript
// Starts your Webhook with default configuration where the Webex Teams API access token is read from the ACCESS_TOKEN env variable 
SparkBot = require("node-sparkbot");
var bot = new SparkBot();
bot.secret = "not THAT secret"
...
```

Note that it is a HIGHLY RECOMMENDED however not mandatory security practice to set up a SECRET. 
If your bot has been started with a secret, then the processing will abort if the incoming payload signature is not present or do not fit.
However, the bot framework defines a flag so that you can ignore signature check failures when a SECRET is defined.


### Auto-Register Webhooks

At startup, node-sparkbot can automatically create a Webhook for your bot, or verify if a webhook already exists with the specified name.
Here are a few of the options supported:
* Simplissime registration where defaults apply (all, all, no filter, no secret), and no callback
   > bot.createOrUpdateWebhook("register-bot", "https://f6d5d937.ngrok.io");
* Registration with no filter, no secret, and no callback
   > bot.createOrUpdateWebhook("register-bot", "https://f6d5d937.ngrok.io", "all", "all");

CHANGELOG

v2.3 - support for the API emulator
- migrated from 'https' to 'got'

v2.2 - new WEBEX_API env variable
- support for an alternative HTTPS endpoint 
- examples: api emulator on Heroku, Glitch or over ngrok

v2.1 - official support for webhook automation
- robustified

v2.0 - Support for Adaptive Cards
- support for AttachmentActions

v1.0 - Webex Teams rebrand
- Added support for 'webex.bot' domain
- Added support for ACCESS_TOKEN
- SPARK_TOKEN deprecated, but still supported
- Documentation updates (removed Spark mentions)

v0.12 - Legacy Cisco Spark release


sparkbot/webhook.js
//
// Copyright (c) 2016-2019 Cisco Systems
// Licensed under the MIT License 
//

var express = require("express");
var app = express();
var bodyParser = require("body-parser");
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

var debug = require("debug")("sparkbot");
var fine = require("debug")("sparkbot:fine");

var Utils = require("./utils");
var CommandInterpreter = require("./interpreter");
var CommandRouter = require("./router");
var Registration = require("./registration");


/* Creates a Webex Teams webhook with specified configuration structure: 
 *  
 *  { 
 * 		port,				// int: local port on which the webhook is accessible
 * 		path,				// string: path to which new webhook POST events are expected
 * 		token,				// string: Webex Teams API access token
 * 		secret,				// string: (optional) webhook secret used to sign payloads
 * 		softSecretCheck,	// boolean: does not aborts payload processing if the payload signature check fails
 *		trimMention			// boolean: filters out mentions if token owner is a bot
 * 		ignoreSelf			// boolean: ignores message created by token owner
 *  }
 *  
 * If no configuration is specified, the defaults below apply: 
 *  { 
 *		port				: process.env.PORT || 8080,
 *		path				: process.env.WEBHOOK_PATH || "/" ,
 *		token				: process.env.ACCESS_TOKEN,
 *		secret				: process.env.WEBHOOK_SECRET,
 *		softSecretCheck		: will default to false if a secret is defined,
 *		trimMention			: will default to true,
 *		commandPrefix		: process.env.COMMAND_PREFIX || "/",
 *		ignoreSelf			: will default to true if a bot is used, and false otherwise
 *  }
 * 
 */
function Webhook(config) {
   // Inject defaults if no configuration specified
   if (!config) {
      debug("webhook instantiated with default configuration");
      config = {
         // [WORKAROUND] in some container situation (ie, Cisco Shipped), we need to use an OVERRIDE_PORT to force our bot to start and listen to the port defined in the Dockerfile (ie, EXPOSE), 
         // and not the PORT dynamically assigned by the host or scheduler.
         port: process.env.OVERRIDE_PORT || process.env.PORT,
         path: process.env.WEBHOOK_PATH,
         // [COMPAT] SPARK_TOKEN deprecated but still supported
         token: process.env.ACCESS_TOKEN || process.env.SPARK_TOKEN,
         secret: process.env.WEBHOOK_SECRET,
         commandPrefix: process.env.COMMAND_PREFIX || "/"
      };
   }

   // Robustify: it is usually safer not to read ourselves, especially if no commandPrefix is specified
   if (!config.ignoreSelf && !config.commandPrefix) {
      debug("WARNING: configuration does not prevent for reading from yourself => possible infinite loop, continuing...");
   }

   // Abort if mandatory config parameters are not present
   if (!config.port) {
      fine("no port specified, applying default");
      config.port = 8080;
   }
   if (!config.path) {
      fine("no path specified, applying default");
      config.path = "/";
   }

   // If an access token is specified, create a command interpreter
   if (!config.token) {
      debug("no access token specified, will not fetch message contents and room titles, nor interpret commands");
   }
   this.token = config.token;

   // Apply secret if specified
   if (config.secret) {
      this.secret = config.secret;
      // Defaults to false when a secret is specified
      if (!config.softSecretCheck) {
         this.softSecretCheck = false;
      }
   }
   if (config.softSecretCheck) {
      this.softSecretCheck = config.softSecretCheck;
   }


   // Webhook listeners
   this.listeners = {};
   var self = this;
   function fire(trigger) {
      // Retreive listener for incoming event
      var entry = trigger.resource + "/" + trigger.event;
      var listener = self.listeners[entry];
      if (!listener) {
         debug("no listener found for resource/event: " + entry);
         return;
      }

      // Invoke listener
      debug("calling listener for resource/event: " + entry + ", with data context: " + trigger.data.id);
      listener(trigger);
   }

   // Initialize command processors
   this.interpreter = new CommandInterpreter(config);
   this.router = new CommandRouter(this);

   // Webhook API routes
   started = Date.now();
   app.route(config.path)
      .get(function (req, res) {
         debug("healthcheck hitted");
         var package = require("../package.json");
         res.json({
            message: "Congrats, your bot is up and running",
            since: new Date(started).toISOString(),
            framework: package.name + ", " + package.version,
            tip: "Don't forget to create WebHooks to start receiving events from Webex: https://developer.webex.com/endpoint-webhooks-post.html",
            webhook: {
               secret: (self.secret != null),
               softSecretCheck: self.softSecretCheck,
               listeners: Object.keys(self.listeners),
            },
            token: (self.token != null),
            account: {
               type: self.interpreter.accountType,
               nickName: self.interpreter.nickName,
               person: self.interpreter.person
            },
            interpreter: {
               prefix: self.interpreter.prefix,
               trimMention: self.interpreter.trimMention,
               ignoreSelf: self.interpreter.ignoreSelf
            },
            commands: Object.keys(self.router.commands)
         });
      })
      .post(function (req, res) {
         debug("webhook invoked");

         // analyse incoming payload, should conform to Webex Teams webhook specifications
         if (!req.body || !Utils.checkWebhookEvent(req.body)) {
            debug("unexpected payload POSTed, aborting...");
            res.status(400).json({
               message: "Bad payload for Webhook",
               details: "either the bot is misconfigured or Webex is running a new API version"
            });
            return;
         }

         // event is ready to be processed, let's send a response to Webex without waiting any longer
         res.status(200).json({ message: "notification received and being processed by webhook" });

         // process HMAC-SHA1 signature if a secret has been specified
         // [NOTE@ for security reasons, we check the secret AFTER responding to Webex
         if (self.secret) {
            if (!Utils.checkSignature(self.secret, req)) {
               if (!self.softSecretCheck) {
                  debug("HMAC-SHA1 signature does not match secret, aborting payload processing");
                  return;
               }
               else {
                  fine("HMAC-SHA1 signature does not match secret, continuing....");
               }
            }
            fine("signature check ok, continuing...")
         }

         // process incoming resource/event, see https://developer.webex.com/webhooks-explained.html
         fire(req.body);
      });

   // Start bot
   app.listen(config.port, function () {
      debug("bot started on port: " + config.port);
   }).on('error', (err) => {
      console.log(`cannot launch bot, err: ${err.message}`);
      console.log("existing...")
      process.exit(1);
   });
}


// Registers a listener for new (resource, event) POSTed to our webhook   
Webhook.prototype.onEvent = function (resource, event, listener) {
   if (!listener) {
      debug("on: listener registration error. Please specify a listener for resource/event");
      return;
   }
   // check (resource, event) conforms to Webhook specifications, see https://developer.webex.com/webhooks-explained.html 
   if (!resource || !event) {
      debug("on: listener registration error. please specify a resource/event for listener");
      return;
   }

   switch (resource) {
      case "all":
         if (event != "all") {
            debug("on: listener registration error. Bad configuration: only 'all' events is suported for 'all' resources");
            debug("WARNING: listener not registered for resource/event: " + resource + "/" + event);
            return;
         }

         addAttachmentActionsCreatedListener(this, listener);
         addMessagesCreatedListener(this, listener);
         addMessagesDeletedListener(this, listener);
         addRoomsCreatedListener(this, listener);
         addRoomsUpdatedListener(this, listener);
         addMembershipsCreatedListener(this, listener);
         addMembershipsUpdatedListener(this, listener);
         addMembershipsDeletedListener(this, listener);
         return;

      case "attachmentActions":
         if (event == "all") {
            addAttachmentActionsCreatedListener(this, listener);
            return;
         }
         if (event == "created") {
            addAttachmentActionsCreatedListener(this, listener);
            return;
         }
         break;

      case "messages":
         if (event == "all") {
            addMessagesCreatedListener(this, listener);
            addMessagesDeletedListener(this, listener);
            return;
         }
         if (event == "created") {
            addMessagesCreatedListener(this, listener);
            return;
         }
         if (event == "deleted") {
            addMessagesDeletedListener(this, listener);
            return;
         };
         break;

      case "memberships":
         if (event == "all") {
            addMembershipsCreatedListener(this, listener);
            addMembershipsUpdatedListener(this, listener);
            addMembershipsDeletedListener(this, listener);
            return;
         }
         if (event == "created") {
            addMembershipsCreatedListener(this, listener);
            return;
         }
         if (event == "updated") {
            addMembershipsUpdatedListener(this, listener);
            return;
         };
         if (event == "deleted") {
            addMembershipsDeletedListener(this, listener);
            return;
         };
         break;

      case "rooms":
         if (event == "all") {
            addRoomsCreatedListener(this, listener);
            addRoomsUpdatedListener(this, listener);
            return;
         }
         if (event == "created") {
            addRoomsCreatedListener(this, listener);
            return;
         }
         if (event == "updated") {
            addRoomsUpdatedListener(this, listener);
            return;
         };
         break;

      default:
         break;
   }

   debug("on: listener registration error, bad configuration. Resource: '" + resource + "' and event: '" + event + "' do not comply with Webex Teams webhook specifications.");
}


// Helper function to retreive message details from a (messages/created) Webhook trigger.
// Expected callback function signature (err, message).
Webhook.prototype.decryptMessage = function (trigger, cb) {
   if (!this.token) {
      debug("no access token configured, cannot read message details.")
      cb(new Error("no access token configured, cannot decrypt message"), null);
      return;
   }

   Utils.readMessage(trigger.data.id, this.token, cb);
}


// Utility function to be notified only as new messages are posted into spaces against which your Webhook has registered to.
// The callback function will directly receive the message contents : combines .on('messages', 'created', ...) and .decryptMessage(...).
// Expects a callback function with signature (err, trigger, message).
// Returns true or false whether registration was successful
Webhook.prototype.onMessage = function (cb) {

   // check args
   if (!cb) {
      debug("no callback function, aborting callback registration...")
      return false;
   }

   // Abort if webhook cannot request Webex for messages details
   var token = this.token;
   if (!token) {
      debug("no access token specified, will not read message details, aborting callback registration...")
      return false;
   }

   addMessagesCreatedListener(this, function (trigger) {
      Utils.readMessage(trigger.data.id, token, function (err, message) {
         if (err) {
            debug("could not fetch message details, err: " + JSON.stringify(err) + ", listener not fired...");
            //cb (err, trigger, null);
            return;
         }

         // Fire listener
         cb(trigger, message);
      });
   });

   return true;
}


// Transforms a message into a Command structure
Webhook.prototype.asCommand = function (message) {
   if (!message) {
      debug("no message to interpret, aborting...")
      return null;
   }

   return this.interpreter.extract(message);
}



// Shortcut to be notified only as new commands are posted into spaces your Webhook has registered against.
// The callback function will directly receive the message contents : combines .on('messages', 'created', ...)  .decryptMessage(...) and .extractCommand(...).
// The expected callback function signature is: function(err, command).
// Note that you may register a "fallback" listener by registering the "fallback" command
Webhook.prototype.onCommand = function (command, cb) {
   if (!command || !cb) {
      debug("wrong arguments for .onCommand, aborting...")
      return;
   }

   this.router.addCommand(command, cb);
}


// Utility function to be notified only as new cards are submitted
// The callback function will directly receive the submitted data contents : combines .on('attachmentActions', 'created', ...) and .decryptMessage(...).
// Expects a callback function with signature (err, trigger, submission).
// Returns true or false whether registration was successful
Webhook.prototype.onCardSubmission = function (cb) {

   // check args
   if (!cb) {
      debug("no callback function, aborting callback registration...")
      return false;
   }

   // Abort if webhook cannot request Webex for messages details
   var token = this.token;
   if (!token) {
      debug("no access token specified, will not be able to read submitted details, aborting callback registration...")
      return false;
   }

   addAttachmentActionsCreatedListener(this, function (trigger) {
      Utils.readAttachmentActions(trigger.data.id, token, function (err, attachmentActions) {
         if (err) {
            debug("could not fetch attachmentActions details, err: " + JSON.stringify(err) + ", listener not fired...");
            //cb (err, trigger, null);
            return;
         }

         // Fire listener
         cb(trigger, attachmentActions);
      });
   });

   return true;
}


// Creates or updates a webhook to Webex
// returns the webhook created or updated
// see https://developer.webex.com/endpoint-webhooks-post.html for arguments
Webhook.prototype.createOrUpdateWebhook = function (name, targetUrl, resource, event, filter, secret, cb) {
   if (!name || !targetUrl) {
      debug("bad arguments for createOrUpdateWebhook, aborting webhook creation...")
      if (cb) cb(new Error("bad arguments for createOrUpdateWebhook"), null);
      return;
   }

   if (!resource) {
      resource = "all";
      event = "all";
   }
   if (!event) {
      event = "all";
   }

   // Check if a webhook already exists with the same name
   var token = this.token;
   Registration.listWebhooks(token, function (err, webhooks) {
      if (err) {
         debug("could not retreive webhooks, aborting webhook creation or update...");
         if (cb) cb(new Error("could not retreive the list of webhooks, check your token"), null);
         return;
      }

      var webhook = null;
      webhooks.forEach(function (elem) {
         if (elem.name === name) {
            webhook = elem;
         }
      });

      // if found, check if webhook is different
      if (webhook) {
         var identical = compareWebhooks(webhook, name, targetUrl, resource, event, filter, secret);
         if (identical) {
            debug("webhook already exists with same properties, no creation needed");
            if (cb) cb(null, webhook);
            return;
         }

         // delete the webhook that pre-exists
         Registration.deleteWebhook(token, webhook.id, function (err, code) {
            if (err != null) {
               debug("could not delete existing webhook")
               if (cb) cb(new Error('webhook with same name already exists and could not be updated'), null);
               return;
            }

            Registration.createWebhook(token, name, targetUrl, resource, event, filter, secret, function (err, webhook) {
               if (err != null) {
                  debug("could not create webhook")
                  if (cb) cb(new Error('could NOT create webhook'), null);
                  return;
               }

               fine("webhook successfully updated");
               if (cb) cb(null, webhook);
               return;
            });
         });
      }

      // create Webhook
      else {
         Registration.createWebhook(token, name, targetUrl, resource, event, filter, secret, function (err, webhook) {
            if (err != null) {
               debug("could not create webhook")
               if (cb) cb(new Error('could NOT create webhook'), null);
               return;
            }

            fine("webhook successfully created");
            if (cb) cb(null, webhook);
            return;
         });
      }
   });
}


module.exports = Webhook


//
// Internals
//

function addAttachmentActionsCreatedListener(webhook, listener) {
   webhook.listeners["attachmentActions/created"] = listener;
   fine("addAttachmentActionsCreatedListener: listener registered");
}

function addMessagesCreatedListener(webhook, listener) {
   webhook.listeners["messages/created"] = listener;
   fine("addMessagesCreatedListener: listener registered");
}

function addMessagesDeletedListener(webhook, listener) {
   webhook.listeners["messages/deleted"] = listener;
   fine("addMessagesDeletedListener: listener registered");
}

function addRoomsCreatedListener(webhook, listener) {
   webhook.listeners["rooms/created"] = listener;
   fine("addRoomsCreatedListener: listener registered");
}

function addRoomsUpdatedListener(webhook, listener) {
   webhook.listeners["rooms/updated"] = listener;
   fine("addRoomsUpdatedListener: listener registered");
}

function addMembershipsCreatedListener(webhook, listener) {
   webhook.listeners["memberships/created"] = listener;
   fine("addMembershipsCreatedListener: listener registered");
}

function addMembershipsUpdatedListener(webhook, listener) {
   webhook.listeners["memberships/updated"] = listener;
   fine("addMembershipsUpdatedListener: listener registered");
}

function addMembershipsDeletedListener(webhook, listener) {
   webhook.listeners["memberships/deleted"] = listener;
   fine("addMembershipsDeletedListener: listener registered");
}

// returns true if webhooks are identical
function compareWebhooks(webhook, name, targetUrl, resource, event, filter, secret) {
   if ((webhook.name !== name)
      || (webhook.targetUrl !== targetUrl)
      || (webhook.resource !== resource)
      || (webhook.event !== event)) {
      return false;
   }

   // they look pretty identifty, let's check optional fields
   if (filter) {
      if (filter !== webhook.filter) {
         fine("webhook look pretty similar BUT filter is different");
         return false;
      }
   }
   else {
      if (webhook.filter) {
         fine("webhook look pretty similar BUT filter is different");
         return false;
      }
   }

   if (secret) {
      if (secret !== webhook.secret) {
         fine("webhook look pretty similar BUT secret is different");
         return false;
      }
   }
   else {
      if (webhook.secret) {
         fine("webhook look pretty similar BUT secret is different");
         return false;
      }
   }
   return true;
}



